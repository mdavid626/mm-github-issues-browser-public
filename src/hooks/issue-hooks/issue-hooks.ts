import { useQuery, gql } from '@apollo/client';
import { ApolloError } from '@apollo/client/errors';
import { IssueQueryResult, IssuesQueryResult } from '../../types/issue';

const issuesQuery = gql`
  query {
    repository(owner: "facebook", name: "react") {
      issues(last: 20) {
        edges {
          node {
            id
            createdAt
            title
            state
            number
            url
            author {
              login
            }
          }
        }
      }
    }
  }
`;

export const useIssues = (): [
  IssuesQueryResult | undefined,
  boolean,
  ApolloError | undefined
] => {
  const { data, loading, error } = useQuery<IssuesQueryResult>(issuesQuery);
  return [data, loading, error];
};

const issueQuery = gql`
  {
    repository(owner: "facebook", name: "react") {
      issue(number: 26077) {
        id
        createdAt
        title
        state
        number
        stateReason
        url
        bodyText
        author {
          login
        }
      }
    }
  }
`;

export const useIssue = (): [
  IssueQueryResult | undefined,
  boolean,
  ApolloError | undefined
] => {
  const { data, loading, error } = useQuery<IssueQueryResult>(issueQuery);
  return [data, loading, error];
};
