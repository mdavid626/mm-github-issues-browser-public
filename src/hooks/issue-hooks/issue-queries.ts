import { gql } from '@apollo/client';

export const GetIssuesQuery = gql`
  query getIssues($cursor: String, $search: String!) {
    search(query: $search, after: $cursor, type: ISSUE, first: 10) {
      issueCount
      nodes {
        ... on Issue {
          id
          createdAt
          title
          state
          number
          url
          author {
            login
          }
        }
      }
      pageInfo {
        hasNextPage
        endCursor
      }
    }
  }
`;

export const GetIssueQuery = gql`
  query getIssue($issueNumber: Int!) {
    repository(owner: "facebook", name: "react") {
      id
      issue(number: $issueNumber) {
        id
        createdAt
        title
        state
        number
        stateReason
        url
        body
        author {
          login
        }
        comments(first: 3, orderBy: { field: UPDATED_AT, direction: DESC }) {
          nodes {
            id
            createdAt
            body
            author {
              login
            }
          }
          pageInfo {
            hasNextPage
            endCursor
          }
          totalCount
        }
      }
    }
  }
`;

export const GetIssueCommentsQuery = gql`
  query getIssueComments($issueNumber: Int!, $commentsCursor: String) {
    repository(owner: "facebook", name: "react") {
      id
      issue(number: $issueNumber) {
        id
        comments(
          first: 3
          after: $commentsCursor
          orderBy: { field: UPDATED_AT, direction: DESC }
        ) {
          nodes {
            id
            createdAt
            body
            author {
              login
            }
          }
          pageInfo {
            hasNextPage
            endCursor
          }
          totalCount
        }
      }
    }
  }
`;
