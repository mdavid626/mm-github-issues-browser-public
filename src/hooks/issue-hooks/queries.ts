import { gql } from '@apollo/client';

export const GetIssuesQuery = gql`
  query getIssues($cursor: String, $states: [IssueState!]) {
    repository(owner: "facebook", name: "react") {
      id
      issues(
        first: 10
        after: $cursor
        states: $states
        orderBy: { field: CREATED_AT, direction: DESC }
      ) {
        nodes {
          id
          createdAt
          title
          state
          number
          url
          author {
            login
          }
        }
        pageInfo {
          hasNextPage
          startCursor
          endCursor
        }
        totalCount
      }
    }
  }
`;

export const GetIssueQuery = gql`
  query getIssue($issueNumber: Int!) {
    repository(owner: "facebook", name: "react") {
      id
      issue(number: $issueNumber) {
        id
        createdAt
        title
        state
        number
        stateReason
        url
        bodyText
        author {
          login
        }
        comments(first: 3, orderBy: { field: UPDATED_AT, direction: DESC }) {
          nodes {
            id
            createdAt
            bodyText
            author {
              login
            }
          }
          pageInfo {
            hasPreviousPage
            hasNextPage
            startCursor
            endCursor
          }
          totalCount
        }
      }
    }
  }
`;

export const GetCommentsQuery = gql`
  query getIssueComments($issueNumber: Int!, $commentsCursor: String) {
    repository(owner: "facebook", name: "react") {
      id
      issue(number: $issueNumber) {
        id
        comments(
          first: 3
          after: $commentsCursor
          orderBy: { field: UPDATED_AT, direction: DESC }
        ) {
          nodes {
            id
            createdAt
            bodyText
            author {
              login
            }
          }
          pageInfo {
            hasPreviousPage
            hasNextPage
            startCursor
            endCursor
          }
          totalCount
        }
      }
    }
  }
`;
