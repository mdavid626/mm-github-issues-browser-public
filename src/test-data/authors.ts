import { Author } from '../types/author';

export const testAuthor1: Author = {
  login: 'testAuthor1',
};
