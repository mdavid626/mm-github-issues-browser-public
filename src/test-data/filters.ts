import { Filters } from '../types/filters';

export const testFilters1: Filters = {
  search: 'test',
  state: 'open',
};
