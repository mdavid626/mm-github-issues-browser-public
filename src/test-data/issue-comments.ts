import { IssueComment, IssueComments } from '../types/issue-comment';
import { testAuthor1 } from './authors';

export const testIssueComment1: IssueComment = {
  id: 'IC_kwDOAJy2Ks5Ux_nZ',
  body: 'issue comment 1',
  createdAt: '2023-02-12T09:43:24.641Z',
  author: testAuthor1,
};

export const testIssueComment2: IssueComment = {
  id: 'IC_kwDOAJy2Ks5Ux4u7',
  body: 'issue comment 2',
  createdAt: '2022-01-11T10:42:25.655Z',
  author: testAuthor1,
};

export const testIssueComments1: IssueComments = {
  nodes: [testIssueComment1, testIssueComment2],
  pageInfo: {
    hasNextPage: true,
    endCursor: 'endCursor1',
  },
  totalCount: 2,
};

export const testEmptyIssueComments: IssueComments = {
  nodes: [],
  pageInfo: {
    hasNextPage: false,
    endCursor: 'endCursor1',
  },
  totalCount: 0,
};
