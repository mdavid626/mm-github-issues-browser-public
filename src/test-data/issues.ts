import { Issue } from '../types/issue';
import { testAuthor1 } from './authors';
import { testIssueComment1, testIssueComment2 } from './issue-comments';

export const testIssue1: Issue = {
  id: 'I_kwDOAJy2Ks5d65oG',
  createdAt: '2023-02-08T08:50:14Z',
  title: 'Bug: Failed when set homepage in package.json or set PUBLIC_URL',
  state: 'CLOSED',
  number: 26126,
  url: 'https://github.com/facebook/react/issues/26126',
  body: 'test issue1 body',
  author: testAuthor1,
  comments: {
    nodes: [testIssueComment1, testIssueComment2],
    pageInfo: {
      hasNextPage: true,
      endCursor: 'endCursor1',
    },
    totalCount: 2,
  },
};

export const noCommentsIssue1: Issue = {
  id: 'IC_kwDOAJy2Ks5Ux_nZ',
  createdAt: '2022-02-08T08:50:14Z',
  title: 'no comments issue1',
  state: 'CLOSED',
  number: 26127,
  url: 'https://github.com/facebook/react/issues/26127',
  body: 'no comments issue body',
  author: testAuthor1,
  comments: {
    nodes: [],
    pageInfo: {
      hasNextPage: false,
      endCursor: 'endCursor1',
    },
    totalCount: 0,
  },
};
