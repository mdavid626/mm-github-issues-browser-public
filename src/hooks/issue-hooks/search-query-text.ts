import { Filters } from '../../types/filters';

const createSearchQueryText = (filters: Filters) => {
  const escapedSearch = filters.search.replace(/[^a-z\d\s]/gi, '');
  const search = filters.search
    ? `(in:title ${escapedSearch} OR in:body ${escapedSearch})`
    : '';
  const state = filters.state ? `state:${filters.state}` : '';
  return ['repo:facebook/react', 'type:issue', state, search]
    .filter((item) => item)
    .join(' ');
};

export default createSearchQueryText;
