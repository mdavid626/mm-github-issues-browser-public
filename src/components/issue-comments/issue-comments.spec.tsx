import { cleanup, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';
import { useFetchMoreIssueComment } from '../../hooks/issue-hooks/issue-hooks';
import { noCommentsIssue1, testIssue1 } from '../../test-data/issues';
import IssueCommentItem from '../issue-comment-item/issue-comment-item';
import IssueComments from './issue-comments';

jest.mock('../../hooks/issue-hooks/issue-hooks');
jest.mock('../issue-comment-item/issue-comment-item');

describe('issue-comments', () => {
  beforeEach(() => {
    (IssueCommentItem as jest.Mock).mockReturnValue(
      <div>issue-comment-item</div>
    );
    (useFetchMoreIssueComment as jest.Mock).mockReturnValue([jest.fn(), false]);
  });
  afterEach(cleanup);
  afterEach(jest.resetAllMocks);

  it('should render', () => {
    const { asFragment } = render(
      <IssueComments
        issueNumber={testIssue1.number}
        comments={testIssue1.comments}
      />
    );
    expect(asFragment()).toMatchSnapshot();
    expect(useFetchMoreIssueComment).toHaveBeenCalledWith(testIssue1.number);
    expect(IssueCommentItem).toHaveBeenCalledTimes(2);
    expect(IssueCommentItem).toHaveBeenCalledWith(
      { comment: testIssue1.comments.nodes[0] },
      {}
    );
    expect(IssueCommentItem).toHaveBeenCalledWith(
      { comment: testIssue1.comments.nodes[1] },
      {}
    );
  });

  it('should render no comments message when no comments', () => {
    render(
      <IssueComments
        issueNumber={noCommentsIssue1.number}
        comments={noCommentsIssue1.comments}
      />
    );
    expect(screen.getByText('No comments yet')).toBeVisible();
    expect(IssueCommentItem).not.toHaveBeenCalled();
  });

  it('should fetch more comments when button clicked', async () => {
    const fetchMoreIssueComment = jest.fn();
    (useFetchMoreIssueComment as jest.Mock).mockReturnValue([
      fetchMoreIssueComment,
      false,
    ]);
    render(
      <IssueComments
        issueNumber={testIssue1.number}
        comments={testIssue1.comments}
      />
    );
    await userEvent.click(screen.getByText('load more comments'));
    expect(fetchMoreIssueComment).toHaveBeenCalledWith(
      testIssue1.comments.pageInfo.endCursor
    );
  });

  it('should show loading spinner when loading more comments', () => {
    (useFetchMoreIssueComment as jest.Mock).mockReturnValue([jest.fn(), true]);
    render(
      <IssueComments
        issueNumber={testIssue1.number}
        comments={testIssue1.comments}
      />
    );
    expect(screen.getByText('loading...')).toBeVisible();
    expect(screen.queryByText('load more comments')).toBe(null);
  });
});
