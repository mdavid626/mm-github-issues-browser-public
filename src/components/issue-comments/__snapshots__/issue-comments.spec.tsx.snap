// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`issue-comments should render 1`] = `
<DocumentFragment>
  <div
    class="IssueComments"
  >
    <div
      class="IssueComments-title"
    >
      Comments (2)
    </div>
    <div
      class="IssueComments-comments"
    >
      <div>
        issue-comment-item
      </div>
      <div>
        issue-comment-item
      </div>
    </div>
    <div
      class="IssueComments-loadMore"
    >
      load more comments
    </div>
  </div>
</DocumentFragment>
`;
